File: Example/ZLForm/ZLFormTextFieldCell.m
//
//  ZLFormTextFieldCell.m
//  ZLForm
//
//  Created by admin on 2026/5/20.
//

#import "ZLFormTextFieldCell.h"
#if __has_include(<ZLForm/ZLForm.h>)
#import <ZLForm/ZLForm.h>
#else
#import "ZLForm.h"
#endif

@interface ZLFormTextFieldCell ()<UITextFieldDelegate>
@property (nonatomic, strong, readwrite) UITextField *textField;
@end

@implementation ZLFormTextFieldCell

- (void)configure {
    [super configure];
    self.selectionStyle = UITableViewCellSelectionStyleNone;
    
    self.textField = [[UITextField alloc] init];
    self.textField.translatesAutoresizingMaskIntoConstraints = NO;
    self.textField.textAlignment = NSTextAlignmentRight;
    self.textField.font = [UIFont systemFontOfSize:15];
    self.textField.textColor = [UIColor darkTextColor];
    self.textField.returnKeyType = UIReturnKeyDone;
    self.textField.delegate = self;
    [self.textField addTarget:self action:@selector(textFieldDidChange:) forControlEvents:UIControlEventEditingChanged];
    [self.contentView addSubview:self.textField];
    
    // Layout: titleLabel on left, textField on right
    [NSLayoutConstraint activateConstraints:@[
        [self.titleLabel.leadingAnchor constraintEqualToAnchor:self.contentView.leadingAnchor constant:15],
        [self.titleLabel.centerYAnchor constraintEqualToAnchor:self.contentView.centerYAnchor],
        [self.titleLabel.widthAnchor constraintEqualToConstant:80],
        
        [self.textField.leadingAnchor constraintEqualToAnchor:self.titleLabel.trailingAnchor constant:10],
        [self.textField.trailingAnchor constraintEqualToAnchor:self.contentView.trailingAnchor constant:-15],
        [self.textField.centerYAnchor constraintEqualToAnchor:self.contentView.centerYAnchor],
        [self.textField.heightAnchor constraintEqualToConstant:36],
    ]];
    
    // Hide the detailLabel from base cell
    self.detailLabel.hidden = YES;
}

- (void)update {
    [super update];
    self.titleLabel.text = self.rowDescriptor.title;
    self.textField.text = [self.rowDescriptor valueForDisplay];
    self.textField.placeholder = self.rowDescriptor.emptyDisplayValue;
    self.textField.enabled = !self.rowDescriptor.disabled;
}

#pragma mark - UITextFieldDelegate

- (void)textFieldDidChange:(UITextField *)textField {
    id oldValue = self.rowDescriptor.value;
    self.rowDescriptor.value = textField.text;
    if (self.rowDescriptor.onChangeBlock) {
        self.rowDescriptor.onChangeBlock(oldValue, textField.text, self.rowDescriptor);
    }
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    [textField resignFirstResponder];
    return YES;
}

@end

File: Example/ZLForm/ZLFormTextFieldCell.swift
import UIKit
import ZLForm
import SnapKit

@objcMembers
public class ZLFormTextFieldCell: ZLFormBaseCell, UITextFieldDelegate {
    public private(set) var textField: UITextField!
    @objc public private(set) var titleLabel: UILabel!
    
    open override func configure() {
        super.configure()
        selectionStyle = .none
        
        textField = UITextField()
        textField.textAlignment = .right
        textField.font = UIFont.systemFont(ofSize: 15)
        textField.textColor = .darkText
        textField.returnKeyType = .done
        textField.delegate = self
        textField.addTarget(self, action: #selector(textFieldDidChange(_:)), for: .editingChanged)
        contentView.addSubview(textField)
        
        titleLabel = UILabel()
        titleLabel.numberOfLines = 0
        titleLabel.font = UIFont.systemFont(ofSize: 16)
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        titleLabel.setContentCompressionResistancePriority(.defaultHigh + 1, for: .horizontal)
        contentView.addSubview(titleLabel)
        
        
        titleLabel.snp.remakeConstraints { make in
            make.leading.equalToSuperview().offset(15)
            make.centerY.equalToSuperview()
            make.width.equalTo(80)
        }
        
        textField.snp.makeConstraints { make in
            make.leading.equalTo(titleLabel.snp.trailing).offset(10)
            make.trailing.equalToSuperview().offset(-15)
            make.centerY.equalToSuperview()
            make.height.equalTo(36)
        }
    }
    public override func setupSubviews() {
        
    }
    
    open override func update() {
        super.update()
        guard let row = rowDescriptor else { return }
        titleLabel.text = row.title
        textField.text = row.valueForDisplay() as? String;
        textField.placeholder = row.placeholderValue as? String
        textField.isEnabled = !row.disabled
    }
    
    // MARK: - UITextFieldDelegate
    
    @objc private func textFieldDidChange(_ textField: UITextField) {
        let oldValue = rowDescriptor?.value
        rowDescriptor?.value = textField.text
        rowDescriptor?.onChangeBlock?(oldValue, textField.text, rowDescriptor!)
    }
    
    public func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return true
    }
}
